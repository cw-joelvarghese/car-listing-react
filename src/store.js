import { configureStore } from "@reduxjs/toolkit";
import carsReducers from "./reducers/carsReducer";
import filterReducer from "./reducers/filterReducer";

export default configureStore({
    reducer: {
        cars: carsReducers,
        filter: filterReducer
    },
});
