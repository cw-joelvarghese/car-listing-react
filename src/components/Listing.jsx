import { useContext, useEffect, useState } from "react";
import styles from "./Listing.module.css";
import { CarContext } from "src/App";

function CarCard({ car }) {
    // priceNumeric
    return (
        <div className={styles.carCard}>
            <img src={car.imageUrl} className={styles.carImage} />
            <div className={styles.carInfo}>
                <div>
                    <h3>
                        {car.makeYear} {car.makeName}
                    </h3>
                    <p>{car.km} km | {car.fuel} | {car.areaName}, {car.cityName}</p>
                </div>

                <div className={styles.carEMIAndPrice}>
                    <p>Rs. {car.price}</p>
                    <p>{car.emiText	}</p>
                </div>
                <p>Make Offer</p>
                <button className={styles.carSellerButton}>
                    Get Seller info
                </button>
            </div>
        </div>
    );
}

function Listing() {
    const { cars } = useContext(CarContext);
    return (
        <div className={styles.container}>
            <div className={styles.filterContainer}>
                <p>Sort by: </p>
                <select>
                    <option>Price: Low to high</option>
                    <option>Price: High to low</option>
                </select>
            </div>
            
            {cars?.map((car) => 
                <CarCard key={car.profileId	} car={car} />
            )}
        </div>
    );
}

export default Listing;
