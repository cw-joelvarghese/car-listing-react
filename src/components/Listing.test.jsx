import { render, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import Listing, { CarCard } from "src/components/Listing";
import store from "src/store";
import { expect, test, vi } from "vitest";
import '@testing-library/jest-dom'


const car = {
    imageUrl: "https://image.com",
    makeYear: 2002,
    makeName: "Suzuki",
    km: 200,
    fuel: 1,
    areaName: "Worli",
    cityName: "Mumbai",
    price: 12,
    emiText: "Starts from Rs. 12 Lakh",
    profileId: 1,
};

global.fetch = vi.fn();

// Creating a fetch replacement function with .json() implementation
function createFetchResponse(data) {
    return { json: () => new Promise((resolve) => resolve(data)) };
}

test("Listing component renders correctly", async () => {
    const response = {
        stocks: [
            {
                imageUrl: "https://image.com",
                makeYear: 2002,
                makeName: "Suzuki",
                km: 200,
                fuel: 1,
                areaName: "Worli",
                cityName: "Mumbai",
                price: 12,
                emiText: "Starts from Rs. 12 Lakh",
                profileId: 1,
            },
            {
                imageUrl: "https://image.com",
                makeYear: 2002,
                makeName: "Mazda",
                km: 200,
                fuel: 1,
                areaName: "Worli",
                cityName: "Mumbai",
                price: 12,
                emiText: "Starts from Rs. 12 Lakh",
                profileId: 2,
            },
        ],
    };

    // Mocking request with fixed data
    fetch.mockResolvedValue(createFetchResponse(response));
    const tree = render(
        <Provider store={store}>
            <Listing />
        </Provider>
    );
    await waitFor(() => {
        expect(tree.getByText(/Suzuki/i)).toBeInTheDocument()
        expect(tree).toMatchSnapshot();
    });
});

test("CarCard component renders correctly", () => {
    const tree = render(<CarCard car={car} />);
    expect(tree).toMatchSnapshot();
});
