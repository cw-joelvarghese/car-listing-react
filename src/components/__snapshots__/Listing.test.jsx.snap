// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CarCard component renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_carCard_187b3e"
      >
        <img
          class="_carImage_187b3e"
          src="https://image.com"
        />
        <div
          class="_carInfo_187b3e"
        >
          <div>
            <h3>
              2002
               
              Suzuki
            </h3>
            <p>
              200
               km | 
              1
               | 
              Worli
              ,
               
              Mumbai
            </p>
          </div>
          <div
            class="_carEMIAndPrice_187b3e"
          >
            <p>
              Rs. 
              12
            </p>
            <p>
              Starts from Rs. 12 Lakh
            </p>
          </div>
          <p>
            Make Offer
          </p>
          <button
            class="_carSellerButton_187b3e"
          >
            Get Seller info
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_carCard_187b3e"
    >
      <img
        class="_carImage_187b3e"
        src="https://image.com"
      />
      <div
        class="_carInfo_187b3e"
      >
        <div>
          <h3>
            2002
             
            Suzuki
          </h3>
          <p>
            200
             km | 
            1
             | 
            Worli
            ,
             
            Mumbai
          </p>
        </div>
        <div
          class="_carEMIAndPrice_187b3e"
        >
          <p>
            Rs. 
            12
          </p>
          <p>
            Starts from Rs. 12 Lakh
          </p>
        </div>
        <p>
          Make Offer
        </p>
        <button
          class="_carSellerButton_187b3e"
        >
          Get Seller info
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Listing component renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_187b3e"
      >
        <div
          class="_filterContainer_187b3e"
        >
          <p>
            Sort by: 
          </p>
          <select>
            <option
              value="1"
            >
              Unsorted
            </option>
            <option
              value="2"
            >
              Price: Low to high
            </option>
            <option
              value="3"
            >
              Price: High to low
            </option>
          </select>
        </div>
        <div
          class="_carCard_187b3e"
        >
          <img
            class="_carImage_187b3e"
            src="https://image.com"
          />
          <div
            class="_carInfo_187b3e"
          >
            <div>
              <h3>
                2002
                 
                Suzuki
              </h3>
              <p>
                200
                 km | 
                1
                 | 
                Worli
                ,
                 
                Mumbai
              </p>
            </div>
            <div
              class="_carEMIAndPrice_187b3e"
            >
              <p>
                Rs. 
                12
              </p>
              <p>
                Starts from Rs. 12 Lakh
              </p>
            </div>
            <p>
              Make Offer
            </p>
            <button
              class="_carSellerButton_187b3e"
            >
              Get Seller info
            </button>
          </div>
        </div>
        <div
          class="_carCard_187b3e"
        >
          <img
            class="_carImage_187b3e"
            src="https://image.com"
          />
          <div
            class="_carInfo_187b3e"
          >
            <div>
              <h3>
                2002
                 
                Mazda
              </h3>
              <p>
                200
                 km | 
                1
                 | 
                Worli
                ,
                 
                Mumbai
              </p>
            </div>
            <div
              class="_carEMIAndPrice_187b3e"
            >
              <p>
                Rs. 
                12
              </p>
              <p>
                Starts from Rs. 12 Lakh
              </p>
            </div>
            <p>
              Make Offer
            </p>
            <button
              class="_carSellerButton_187b3e"
            >
              Get Seller info
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_187b3e"
    >
      <div
        class="_filterContainer_187b3e"
      >
        <p>
          Sort by: 
        </p>
        <select>
          <option
            value="1"
          >
            Unsorted
          </option>
          <option
            value="2"
          >
            Price: Low to high
          </option>
          <option
            value="3"
          >
            Price: High to low
          </option>
        </select>
      </div>
      <div
        class="_carCard_187b3e"
      >
        <img
          class="_carImage_187b3e"
          src="https://image.com"
        />
        <div
          class="_carInfo_187b3e"
        >
          <div>
            <h3>
              2002
               
              Suzuki
            </h3>
            <p>
              200
               km | 
              1
               | 
              Worli
              ,
               
              Mumbai
            </p>
          </div>
          <div
            class="_carEMIAndPrice_187b3e"
          >
            <p>
              Rs. 
              12
            </p>
            <p>
              Starts from Rs. 12 Lakh
            </p>
          </div>
          <p>
            Make Offer
          </p>
          <button
            class="_carSellerButton_187b3e"
          >
            Get Seller info
          </button>
        </div>
      </div>
      <div
        class="_carCard_187b3e"
      >
        <img
          class="_carImage_187b3e"
          src="https://image.com"
        />
        <div
          class="_carInfo_187b3e"
        >
          <div>
            <h3>
              2002
               
              Mazda
            </h3>
            <p>
              200
               km | 
              1
               | 
              Worli
              ,
               
              Mumbai
            </p>
          </div>
          <div
            class="_carEMIAndPrice_187b3e"
          >
            <p>
              Rs. 
              12
            </p>
            <p>
              Starts from Rs. 12 Lakh
            </p>
          </div>
          <p>
            Make Offer
          </p>
          <button
            class="_carSellerButton_187b3e"
          >
            Get Seller info
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
